webapp/api/launchpad.py

```diff
@@ -1,6 +1,105 @@
-from launchpadlib.launchpad import Launchpad
+from hashlib import md5
+from webapp.api.requests import Session
 
 
-launchpad = Launchpad.login_anonymously(
-    "storefront", "production", version="devel"
-)
+class Launchpad:
+    """
+    Provides authentication for Launchpad.
+    """
+
+    BASE_URL = "https://api.launchpad.net"
+
+    def __init__(self, username, token, signature, session=Session()):
+        self.username = username
+        self.session = session
+        self.session.headers["Accept"] = "application/json"
+        self.session.headers["Authorization"] = (
+            f'OAuth oauth_version="1.0", '
+            f'oauth_signature_method="PLAINTEXT", '
+            f"oauth_consumer_key={username}, "
+            f'oauth_token="{token}", '
+            f'oauth_signature="&{signature}"'
+        )
+
+    def _request(self, method, url, params={}, data={}):
+        """
+        Makes a raw HTTP request and returns the response.
+        """
+        if not url.startswith("http"):
+            url = f"{self.BASE_URL}/devel/{url}"
+
+        response = self.session.request(method, url, params=params, data=data)
+        response.raise_for_status()
+
+        return response
+
+    def get(self, resource, params=None):
+        """
+        Makes a raw HTTP GET request and returns the response.
+        """
+        response = self._request("GET", resource, params=params)
+
+        return response.json()
+
+    def post(self, resource, data=None):
+        """
+        Makes a raw HTTP POST request and returns the response.
+        """
+        response = self._request("POST", resource, data=data)
+
+        return response.json() if response.content else response.ok
+
+    def get_collection_entries(self, resource, params=None):
+        """
+        Return collection items from the API
+        """
+        collection = self.get(resource, params)
+
+        return collection.get("entries", [])
+
+    def get_snap_by_store_name(self, snap_name):
+        """
+        Return an Snap from the Launchpad API by store_name
+        """
+        snaps = self.get_collection_entries(
+            "+snaps",
+            {
+                "ws.op": "findByStoreName",
+                "owner": "/~build.snapcraft.io",
+                "store_name": snap_name,
+            },
+        )
+
+        # The Launchpad API only allow to perform a find by store_name
+        # but we are only interested in exactly this one
+        if snaps and snaps[0]["store_name"] == snap_name:
+            return snaps[0]
+
+        return None
+
+    def new_snap(self, snap_name, git_url):
+        """
+        Create an ISnap in Launchpad
+        """
+        data = {
+            "ws.op": "new",
+            "owner": f"/~{self.username}",
+            "name": md5(snap_name.encode("UTF-8")).hexdigest(),
```

Comment:
This is kinda weird. Is this basically a unique ID?

Reply:
Yes, for this API we use always the same user, it is the same user that build.snapcraft.io is using. All the launchpad snaps created with build.s.i are owned by this user. Let me know when you need the environment variables.

Reply:
User? `md5(snap_name.encode("UTF-8")).hexdigest()` presumably generates a hash that's based on the snap name rather than the user's name doesn't it?

Reply:
Do you mean the reason for the "/\~" before the username? It's a pattern that launchpad follow I don't know exactly how it works, they have "~" and "+" for the different entry types. At the beginning, while I was testing with this Colin told me to use this prefix for the username and "+" for the snaps for the endpoint. We are doing the same thing currently in build.s.i:
https://github.com/canonical-web-and-design/build.snapcraft.io/blob/f3c648a1f463efc10f64bfea1b8eabb4965844e6/src/server/handlers/launchpad.js#L200

Reply:
I'm talking about line 87, not line 86

Reply:
Oh wait, there is an issue in that line, it should be the md5 of the git repo, not the snap name.

For the name property in launchpad we are using the md5 of the git repo and the actual snap name is on "store_name". I am following this pattern from build.s.i. I didn't ask why.

Reply:
Yeah that's fine I guess. It would be nice to know why, but probably not worth it here.